lib/spack/spack/test/cmd/location.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import shutil
+import pytest
+
+from llnl.util.filesystem import mkdirp
+import spack.environment as ev
+from spack.cmd.location import BadEnvironmentError, MissingBuildDirectoryError
+from spack.cmd.location import SpecificationError
+from spack.main import SpackCommand
+import spack.paths
+
+
+# everything here uses the mock_env_path
+pytestmark = pytest.mark.usefixtures('config')
+
+# location prints out "locations of packages and spack directories"
+location = SpackCommand('location')
+
+_repo_dir = os.path.join(spack.paths.repos_path, 'builtin.mock')
+_pkg_dir = os.path.join(_repo_dir, 'packages')
+
+_mock_spec_name = 'externaltest'
+_test_env_name = 'test'
+
+
+@pytest.fixture
+def mock_test_env(mutable_mock_env_path):
+    env_dir = ev.root(_test_env_name)
+    mkdirp(env_dir)
+    yield
+
+    shutil.rmtree(env_dir)
+
+
+@pytest.fixture
+def mock_spec():
+    s = spack.spec.Spec(_mock_spec_name).concretized()
+    pkg = spack.repo.get(s)
+
+    # Make it look like the source was actually expanded.
+    source_path = pkg.stage.source_path
+    mkdirp(source_path)
+    yield
+
+    # The test is done so be sure to remove the spec from the mock stage area.
+    shutil.rmtree(pkg.stage.path)
+
+
+def test_location_bad_env():
+    """Tests spack location with an unknown environment."""
+    with pytest.raises(BadEnvironmentError, match="No such environment"):
+        location('--env', 'test-unknown-env')
+
+
+@pytest.mark.parametrize('spec1,spec2,msg', [
+    (None, None, "You must supply a spec"),
+    ('spec1', 'spec2', "Only one spec, not 2")])
+def test_location_bad_spec_options(spec1, spec2, msg):
```

Comment:
IMO this test would be clearer if there were 2 parameters: the list of specs and the expected message:

```
@pytest.mark.parametrize('specs,msg', [
    ([], "You must supply a spec"),
    (['spec1', 'spec2'], "Only one spec, not 2")])
def test_location_bad_spec_options(specs, msg):
```

I think the parameterization of `test_location_options` could also be simplified in this way (although I found the logic less confusing in that case).

Reply:
Ah.  I think I was missing something syntactically when I first tried to use lists.  Thanks!